src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -0,0 +1,350 @@
+        } finally {
+          bufferLock.unlock();
+        }
+      }
+    }
+  }
+
+  /* TODO: SNOW-529753 Deal with rebalances and restarts */
+  private void init(long recordOffset) {
+    try {
+      startCleaner(recordOffset);
+    } catch (Exception e) {
+      LOGGER.error("Cleaner and Flusher threads shut down before initialization");
+    }
+  }
+
+  // we will deal with restart/rebalance issues after we get the basics working.
+  void startCleaner(long recordOffset) {}
+
+  private boolean shouldConvertContent(final Object content) {
+    return content != null && !(content instanceof SnowflakeRecordContent);
+  }
+
+  private boolean isRecordBroken(final SinkRecord record) {
+    return isContentBroken(record.value()) || isContentBroken(record.key());
+  }
+
+  private boolean isContentBroken(final Object content) {
+    return content != null && ((SnowflakeRecordContent) content).isBroken();
+  }
+
+  private SinkRecord handleNativeRecord(SinkRecord record, boolean isKey) {
+    SnowflakeRecordContent newSFContent;
+    Schema schema = isKey ? record.keySchema() : record.valueSchema();
+    Object content = isKey ? record.key() : record.value();
+    try {
+      newSFContent = new SnowflakeRecordContent(schema, content);
+    } catch (Exception e) {
+      LOGGER.error("Native content parser error:\n{}", e.getMessage());
+      try {
+        // try to serialize this object and send that as broken record
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        ObjectOutputStream os = new ObjectOutputStream(out);
+        os.writeObject(content);
+        newSFContent = new SnowflakeRecordContent(out.toByteArray());
+      } catch (Exception serializeError) {
+        LOGGER.error(
+            "Failed to convert broken native record to byte data:\n{}",
+            serializeError.getMessage());
+        throw e;
+      }
+    }
+    // create new sinkRecord
+    Schema keySchema = isKey ? new SnowflakeJsonSchema() : record.keySchema();
+    Object keyContent = isKey ? newSFContent : record.key();
+    Schema valueSchema = isKey ? record.valueSchema() : new SnowflakeJsonSchema();
+    Object valueContent = isKey ? record.value() : newSFContent;
+    return new SinkRecord(
+        record.topic(),
+        record.kafkaPartition(),
+        keySchema,
+        keyContent,
+        valueSchema,
+        valueContent,
+        record.kafkaOffset(),
+        record.timestamp(),
+        record.timestampType(),
+        record.headers());
+  }
+
+  public void insertBufferedRows() {
+    PartitionBuffer intermediateBuffer = null;
+    bufferLock.lock();
+    try {
+      intermediateBuffer = streamingBuffer;
+      streamingBuffer = new StreamingBuffer();
+    } finally {
+      // release lock
+      bufferLock.unlock();
+    }
+    // intermediate buffer can be empty here if time interval reached but kafka produced no records.
+    if (intermediateBuffer.isEmpty()) {
```

Comment:
a small optimization is that we could check the streamingBuffer before taking the lock, and skip if it's empty. I know there could be insert rows after the check, but I assume those rows could be flushed later.